src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs

```diff
@@ -0,0 +1,264 @@
+                                    Debug.Assert(refKind != RefKind.None);
+                                    // A return type can only have one '{0}' modifier.
+                                    diagnostics.Add(ErrorCode.ERR_DupReturnTypeMod, modifier.GetLocation(), modifier.Text);
+                                    break;
+
+                                default:
+                                    // '{0}' is not a valid function pointer return type modifier. Valid modifiers are 'ref' and 'ref readonly'.
+                                    diagnostics.Add(ErrorCode.ERR_InvalidFuncPointerReturnTypeModifier, modifier.GetLocation(), modifier.Text);
+                                    break;
+                            }
+                        }
+                    }
+
+                    returnType = typeBinder.BindType(returnTypeParameter.Type, diagnostics, basesBeingResolved);
+
+                    if (returnType.IsVoidType() && refKind != RefKind.None)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_NoVoidHere, returnTypeParameter.GetLocation());
+                    }
+                }
+
+                var signature = new FunctionPointerMethodSymbol(parent, callingConvention, refKind, returnType);
+
+                if (syntax.Parameters.Count > 1)
+                {
+                    signature._lazyParameterSymbols = ParameterHelpers.MakeParameters(typeBinder,
+                        signature,
+                        syntax.Parameters,
+                        arglistToken: out _,
+                        diagnostics,
+                        allowRefOrOut: true,
+                        allowThis: false,
+                        // PROTOTYPE(func-ptr): Custom modifiers for in
+                        addRefReadOnlyModifier: false,
+                        lastIndex: syntax.Parameters.Count - 2,
+                        parsingFunctionPointer: true);
+                }
+
+                return signature;
+            }
+
+            private FunctionPointerMethodSymbol(FunctionPointerTypeSymbol parent, CallingConvention callingConvention, RefKind refKind, TypeWithAnnotations returnType)
+            {
+                _parent = parent;
+                CallingConvention = callingConvention;
+                RefKind = refKind;
+                ReturnTypeWithAnnotations = returnType;
+            }
+
+            internal override CallingConvention CallingConvention { get; }
+            public override bool ReturnsVoid => ReturnTypeWithAnnotations.IsVoidType();
+            public override RefKind RefKind { get; }
+            public override TypeWithAnnotations ReturnTypeWithAnnotations { get; }
+            public override ImmutableArray<ParameterSymbol> Parameters => _lazyParameterSymbols;
+            // PROTOTYPE(func-ptr): Custom modifiers for ref readonly
+            public override ImmutableArray<CustomModifier> RefCustomModifiers => throw new NotImplementedException();
```

Comment:
>throw new NotImplementedException(); [](start = 81, length = 36)

It doesn't feel like throwing is the right thing to do. An empty set would be a better temporary implementation.

Reply:
With the current tests I have, nothing hits this code. I'd rather leave it throwing an exception that will result in an obvious test failure later than make a silent input to a later test that affects code in a way I don't understand.

---
In reply to: [359078224](https://github.com/dotnet/roslyn/pull/40431#discussion_r359078224) [](ancestors = 359078224)

Reply:
> With the current tests I have, nothing hits this code. I'd rather leave it throwing an exception that will result in an obvious test failure later than make a silent input to a later test that affects code in a way I don't understand.

With either approach I think a PROTOTYPE comment is in order. Something should ensure we add tests for the code path.

---
In reply to: [360605306](https://github.com/dotnet/roslyn/pull/40431#discussion_r360605306) [](ancestors = 360605306,359078224)

Reply:
> With either approach I think a PROTOTYPE comment is in order.

That's why there's a prototype comment above it.

---
In reply to: [360612751](https://github.com/dotnet/roslyn/pull/40431#discussion_r360612751) [](ancestors = 360612751,360605306,359078224)